### PR TITLE
introduce TOLERABLE_STATUS_CODES

### PR DESCRIPTION
**Pull request by meyhane**
introduce TOLERABLE_STATUS_CODES so that pipe execution won't fail
revised env vars and service_config_property

### CODE REVIEW

File: service/transform-service.py [hunk 1]
```diff
@@ -89,26 +92,41 @@ def generate(entities, endpoint):
             for index, entity in enumerate(entities):
                 if index > 0:
                     yield ","
-                url_per_entity, method_per_entity, headers_per_entity, prop_per_entity = url, method, headers, prop
+                url_per_entity, method_per_entity, headers_per_entity, prop_per_entity, tolerable_status_codes_per_entity, payload_property_per_entity = url, method, headers, prop, tolerable_status_codes, payload_property
```
Comment (meyhane) [inline service/transform-service.py:95]: 2 new config items are added

File: service/transform-service.py [hunk 2]
```diff
@@ -89,26 +92,41 @@ def generate(entities, endpoint):
...
                     if resp.ok:
-                        entity[prop_per_entity] = resp.json()
+                        transform_result = {"status_code": resp.status_code, "return_value": resp.json()}
                     else:
-                        entity[prop_per_entity] = f'{resp.status_code} - {resp.text}'
+                        transform_result = {"status_code": resp.status_code, "return_value": {"transform_succeeded": resp.ok, "message": resp.text, "status_code": resp.status_code}}
+
+                logger.debug(f'transform of entity with _id={entity.get("_id","?")}, prop_per_entity={prop_per_entity} received {transform_result} from {rendered_url}')
+                if endpoint == 'transform':
+                    if not ((transform_result.get("status_code") >= 200 and transform_result.get("status_code") < 400)
```
Comment (meyhane) [inline service/transform-service.py:123]: if neither error code nor tolerable code abort, otherwise do the expected

File: README.rst [hunk 3]
```diff
@@ -12,35 +12,48 @@ Microservice that calls a URL (with optional payload) and able to store the resu
 * entity level customization
 * transform results are streamed back by default and streaming can be turned off.
 * Listens on port 5001 by default
+* can be configured to tolerate errors - to avoid pump execution failure
+
+Endpoints
```
Comment (meyhane) [inline README.rst:17]: added as it is nice to have

File: README.rst [hunk 4]
```diff
@@ -12,35 +12,48 @@ Microservice that calls a URL (with optional payload) and able to store the resu
...
+  "LOG_LEVEL", "Logging level.", "no", "'INFO'"
+  "PORT", "the port that the service will run on", "no", "5001"
   "DO_STREAM", "Flag to receive responses from this service streamed or not. Streaming will be faster but will always return 200", "no", "true"
   "DO_VERIFY_SSL", "Flag to enable/disable ssl verification", "no", "false"
+  "METHOD *", "http method for the call to URL", "no", "'GET'"
+  "PROPERTY *", "the property that will contain the transformation result", "no", "'response'"
+  "PAYLOAD_PROPERTY_FOR_TRANSFORM_REQUEST *", "the property that will contain the payload to be sent to URL", "no", "'payload'"
+  "URL *", "the url of the system that will provide transformed data", "yes", "n/a"
+  "HEADERS *", "headers in json format for the URL", "no", "n/a"
+  "TOLERABLE_STATUS_CODES *", "regex pattern to be searched on response code from the URL. If matched, error won't be raised but returned( in the field as specified in ""PROPERTY"" variable) so that pipe won't fail. The response will be in this format: {""transform_succeeded"": true|false, ""message"": ""<error message>"", ""status_code"": <http code for error>}. Applicable to `/transform` endpoint only", "no", "none"
```
Comment (meyhane) [inline README.rst:53]: the diff here is big mostly due to reordering items.
it is just lines 54-56 that are new

File: service/transform-service.py (hunk 1 above)
Comment (gabriellsesam) [reply service/transform-service.py:95]: Can't find the declaration of `payload_property`, so I'm guessing `payload_property_per_entity` will be "None" until it's set further down at line 103.

As far as I can tell this won't cause any harm, just wanted to check with you if the naming is intentional as the only other reference to "payload_property" is the removed line 102.

Comment (meyhane) [reply service/transform-service.py:95]: line 21 is where `payload_property` is declared. One of us is looking at the wrong place ,i suppose?

Comment (gabriellsesam) [reply service/transform-service.py:95]: My bad

**Pull request merged**